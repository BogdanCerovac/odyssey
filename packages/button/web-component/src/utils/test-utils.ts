import { AxePuppeteer } from 'axe-puppeteer';
import { toHaveNoViolations } from 'jest-axe';
import { newE2EPage, E2EElement, E2EPage } from '@stencil/core/testing';

export type setupComponentType = Promise<{page: E2EPage, el: E2EElement}>;

export const setupComponent = async function(html: string, selector: string) : setupComponentType {
  const page = await newE2EPage({ html });
  const el = await page.find(selector);
  return {
    page, el
  };
}

/**
 * Performs a11y component validation based on the following tags:
 *    - WCAG 2.0 A && AA
 *    - Section 508
 * @param html
 * @param selector
 */
export const performAccessibilityCheck = async function(html: string, selector: string) {
  const { page } = await setupComponent(html, selector);
  // @ts-ignore
  const results = await new AxePuppeteer(page)
    .include(selector)
    .withTags(['wcag2a', 'wcag2aa', 'section508'])
    .analyze();
  expect(results).toHaveNoViolations();
}

// Jest matcher utils
expect.extend(toHaveNoViolations);

declare global {
  namespace jest {
    interface Matchers<R, T> {
      toHaveNoViolations(): R
    }
  }
}
