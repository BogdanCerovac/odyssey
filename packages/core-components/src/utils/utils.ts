/* eslint-disable import/prefer-default-export */

export type FilterObjectFn = (value: any, key: string | number, obj: object) => boolean;

export const throttle = function (func, wait = 50) {
  let timer = null;
  const self = this;
  return function(...args) {
    if (timer === null) {
      timer = setTimeout(function() {
        func.apply(self, args);
        timer = null;
      }, wait);
    }
  };
};

// modified from okta-ui/packages/courage/src/util/StringUtil.js
// since the string is provided by users, the sprintf does not need to be very strict for error checking
// e.g in o-text-counter-input, we may allow both "Can still input characters" and "Can still input {0} characters"
export const sprintf = function (...params: Array<any>): any {
  const args = Array.prototype.slice.apply(params);
  let value = args.shift();

  for (let i = 0, l = args.length; i < l; i += 1) {
    const entity = args[i];
    value = value.replace('{' + i + '}', entity);
  }

  return value;
};

/**
 * Filter out entries from an object.
 *
 * @param obj - object to filter.
 * @param fn - filter function.
 * @returns a new object without the entries satisfying the filter function.
 */
export function filterObject(obj: object, fn: FilterObjectFn): object {
  return Object.keys(obj).reduce((accum, property) => {
    const value = obj[property];

    if (fn(value, property, obj)) {
      accum[property] = value;
    }

    return accum;
  }, {});
}

/**
 * Check if given parameter is not undefined.
 *
 * @param value - value to check.
 * @returns whether the value is defined.
 */
export function isDefined(value: any): boolean {
  return typeof value !== 'undefined';
}

/**
 * Returns the list of values.
 *
 * @param value - Value or list of values.
 * @returns List of values.
 */
export function getValues(value: string | Array<string>): Array<string> {
  return typeof value !== 'undefined'
    ? [].concat(typeof value === 'string' ? value.split(',') : value)
    : [];
}
