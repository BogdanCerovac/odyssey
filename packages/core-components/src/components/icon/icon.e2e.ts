import { setupComponent, setupComponentType, performAccessibilityCheck } from '../../utils/test-utils';

async function setup(html: string) : setupComponentType {
  return await setupComponent(html, 'o-icon');
}

describe('o-icon', () => {
  describe('Props', () => {
    describe('defaults', () => {
      it('defaults to size=small', async () => {
        const { el } = await setup('<o-icon name="info"></o-icon>');
        expect(el).toMatchSnapshot();
      });
    });

    describe('name', () => {
      it('uses name to generate the className', async () => {
        const { el } = await setup('<o-icon name="warning"></o-icon>');
        expect(el).toMatchSnapshot();
      });
    });

    describe('size', () => {
      it('can be "x-small"', async () => {
        const { el } = await setup('<o-icon name="info" size="x-small"></o-icon>');
        expect(el).toMatchSnapshot();
      });
      it('can be "small"', async () => {
        const { el } = await setup('<o-icon name="info" size="small"></o-icon>');
        expect(el).toMatchSnapshot();
      });
      it('can be "medium"', async () => {
        const { el } = await setup('<o-icon name="info" size="medium"></o-icon>');
        expect(el).toMatchSnapshot();
      });
      it('can be "large"', async () => {
        const { el } = await setup('<o-icon name="info" size="large"></o-icon>');
        expect(el).toMatchSnapshot();
      });
      it('can be "x-large"', async () => {
        const { el } = await setup('<o-icon name="info" size="x-large"></o-icon>');
        expect(el).toMatchSnapshot();
      });
    });
  });

  describe('a11y', () => {
    it('has no accessibility violations', async () => {
      const html = '<o-icon name="warning"></o-icon>';
      await performAccessibilityCheck(html, 'o-icon');
    });
  });
});


