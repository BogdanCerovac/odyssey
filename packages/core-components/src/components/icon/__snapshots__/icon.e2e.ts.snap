// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`o-icon Props defaults defaults to size=small 1`] = `
<o-icon class="hydrated" name="info">
  <mock:shadow-root>
    <span class="o-icon o-icon-info o-icon-size-small" data-se="icon"></span>
  </mock:shadow-root>
</o-icon>
`;

exports[`o-icon Props name uses name to generate the className 1`] = `
<o-icon class="hydrated" name="warning">
  <mock:shadow-root>
    <span class="o-icon o-icon-size-small o-icon-warning" data-se="icon"></span>
  </mock:shadow-root>
</o-icon>
`;

exports[`o-icon Props size can be "large" 1`] = `
<o-icon class="hydrated" name="info" size="large">
  <mock:shadow-root>
    <span class="o-icon o-icon-info o-icon-size-large" data-se="icon"></span>
  </mock:shadow-root>
</o-icon>
`;

exports[`o-icon Props size can be "medium" 1`] = `
<o-icon class="hydrated" name="info" size="medium">
  <mock:shadow-root>
    <span class="o-icon o-icon-info o-icon-size-medium" data-se="icon"></span>
  </mock:shadow-root>
</o-icon>
`;

exports[`o-icon Props size can be "small" 1`] = `
<o-icon class="hydrated" name="info" size="small">
  <mock:shadow-root>
    <span class="o-icon o-icon-info o-icon-size-small" data-se="icon"></span>
  </mock:shadow-root>
</o-icon>
`;

exports[`o-icon Props size can be "x-large" 1`] = `
<o-icon class="hydrated" name="info" size="x-large">
  <mock:shadow-root>
    <span class="o-icon o-icon-info o-icon-size-x-large" data-se="icon"></span>
  </mock:shadow-root>
</o-icon>
`;

exports[`o-icon Props size can be "x-small" 1`] = `
<o-icon class="hydrated" name="info" size="x-small">
  <mock:shadow-root>
    <span class="o-icon o-icon-info o-icon-size-x-small" data-se="icon"></span>
  </mock:shadow-root>
</o-icon>
`;
